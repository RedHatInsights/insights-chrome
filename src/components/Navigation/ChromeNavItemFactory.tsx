import React, { useContext } from 'react';
import { ChromeNavExapandableProps } from './ChromeNavExapandable';
import { ChromeNavGroupProps } from './ChromeNavGroup';
import { ChromeNavItemProps } from './ChromeNavItem';
import { DynamicNavProps } from './DynamicNav';
import NavContext from './navContext';

export type ChromeNavItemFactoryProps = {
  groupId?: string;
  expandable?: boolean;
  dynamicNav?: string;
};

const isNavExpandableProps = (props: Record<string, any>, expandable?: boolean): props is ChromeNavExapandableProps => expandable === true;
const isNavGroupProps = (props: Record<string, any>, groupId?: string): props is ChromeNavGroupProps => typeof groupId === 'string';
const isDynamicNavProps = (props: Record<string, any>, dynamicNav?: string): props is DynamicNavProps => typeof dynamicNav === 'string';

const ChromeNavItemFactory = ({ groupId, expandable, dynamicNav, ...rest }: ChromeNavItemFactoryProps) => {
  const { componentMapper } = useContext(NavContext);
  const props = rest;
  if (isNavGroupProps(props, groupId)) {
    return <componentMapper.group {...props} />;
  } else if (isNavExpandableProps(props, expandable)) {
    return <componentMapper.expandable {...props} />;
  } else if (isDynamicNavProps(props, dynamicNav)) {
    return <componentMapper.dynamicNav {...props} dynamicNav={dynamicNav!} />;
  } else {
    return <componentMapper.item {...(props as ChromeNavItemProps)} />;
  }
};

export default ChromeNavItemFactory;
