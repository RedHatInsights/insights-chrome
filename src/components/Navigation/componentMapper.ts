import ChromeNavGroup from './ChromeNavGroup';
import ChromeNavExapandable from './ChromeNavExapandable';
import ChromeNavItem from './ChromeNavItem';
import DynamicNav from './DynamicNav';

const componentMapper = {
  group: ChromeNavGroup,
  expandable: ChromeNavExapandable,
  item: ChromeNavItem,
  dynamicNav: DynamicNav,
};

export default componentMapper;
