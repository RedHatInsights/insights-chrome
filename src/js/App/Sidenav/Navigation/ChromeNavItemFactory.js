import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import NavContext from './navContext';

const ChromeNavItemFactory = ({ groupId, expandable, ...rest }) => {
  const { componentMapper } = useContext(NavContext);
  let Component;
  if (groupId) {
    Component = componentMapper.group;
  } else if (expandable) {
    Component = componentMapper.expandable;
  } else {
    Component = componentMapper.item;
  }

  return <Component {...rest} />;
};

ChromeNavItemFactory.propTypes = {
  groupId: PropTypes.string,
  expandable: PropTypes.bool,
};

export default ChromeNavItemFactory;
