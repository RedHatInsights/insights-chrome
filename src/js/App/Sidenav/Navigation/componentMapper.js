import ChromeNavGroup from './ChromeNavGroup';
import ChromeNavExapandable from './ChromeNavExapandable';
import ChromeNavItem from './ChromeNavItem';

const componentMapper = {
  group: ChromeNavGroup,
  expandable: ChromeNavExapandable,
  item: ChromeNavItem,
};

export default componentMapper;
