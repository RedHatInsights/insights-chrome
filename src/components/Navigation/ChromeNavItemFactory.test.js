/* eslint-disable react/prop-types */
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import createMockStore from 'redux-mock-store';
import NavContext from './navContext';
import ChromeNavItemFactory from './ChromeNavItemFactory';
import componentMapper from './componentMapper';

const NavContextWrapper = ({
  store,
  providerValue = {
    onLinkClick: jest.fn(),
    componentMapper,
  },
  children,
}) => (
  <MemoryRouter>
    <Provider store={store}>
      <NavContext.Provider value={providerValue}>{children}</NavContext.Provider>
    </Provider>
  </MemoryRouter>
);

describe('ChromeNavItemFactory', () => {
  const mockStore = createMockStore();
  const linkTitle = 'Foo';
  const expandableTitle = 'bar';
  const groupTitle = 'group';
  const itemProps = {
    appId: 'testModule',
    href: '/foo',
    title: linkTitle,
  };
  const expandableItemProps = {
    appId: 'testModule',
    href: '/foo',
    id: 'test-id',
    title: expandableTitle,
    expandable: true,
    routes: [itemProps],
  };
  const groupItemProps = {
    groupId: 'group',
    title: groupTitle,
    navItems: [],
  };
  const store = mockStore({
    chrome: {
      activeModule: 'testModule',
      moduleRoutes: [],
      modules: {
        testModule: {},
      },
    },
  });

  test('should render chrome nav item', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavItemFactory {...itemProps} />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText(linkTitle)).toHaveLength(1);
    expect(queryAllByText(expandableTitle)).toHaveLength(0);
    expect(queryAllByText(groupTitle)).toHaveLength(0);
  });

  test('should render chrome expandable nav item', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavItemFactory {...expandableItemProps} />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText(linkTitle)).toHaveLength(1);
    expect(queryAllByText(expandableTitle)).toHaveLength(1);
    expect(queryAllByText(groupTitle)).toHaveLength(0);
  });

  test('should render chrome group nav item', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavItemFactory {...groupItemProps} />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText(linkTitle)).toHaveLength(0);
    expect(queryAllByText(expandableTitle)).toHaveLength(0);
    expect(queryAllByText(groupTitle)).toHaveLength(1);
  });
});
