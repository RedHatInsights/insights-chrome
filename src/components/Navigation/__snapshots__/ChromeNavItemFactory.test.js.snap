// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChromeNavItemFactory should render chrome expandable nav item 1`] = `
<div>
  <li
    class="pf-c-nav__item pf-m-expandable"
    data-ouia-component-id="OUIA-Generated-NavExpandable-1"
    data-ouia-component-type="PF4/NavExpandable"
    data-ouia-safe="true"
    data-quickstart-id="bar"
  >
    <button
      aria-expanded="false"
      class="pf-c-nav__link"
      id="test-id"
    >
      bar
      <span
        class="pf-c-nav__toggle"
      >
        <span
          class="pf-c-nav__toggle-icon"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 256 512"
            width="1em"
          >
            <path
              d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            />
          </svg>
        </span>
      </span>
    </button>
    <section
      aria-labelledby="test-id"
      class="pf-c-nav__subnav"
      hidden=""
    >
      
      <ul
        class="pf-c-nav__list"
      >
        <li
          class="pf-c-nav__item"
          data-ouia-component-id="Foo"
          data-ouia-component-type="PF4/NavItem"
          data-ouia-safe="true"
        >
          <a
            class="pf-c-nav__link"
            data-quickstart-id="foo"
            data-testid="router-link"
            href="/foo"
          >
            Foo
             
          </a>
        </li>
      </ul>
    </section>
  </li>
</div>
`;

exports[`ChromeNavItemFactory should render chrome group nav item 1`] = `
<div>
  <section
    aria-labelledby="group"
    class="pf-c-nav__section chr-c-section-nav"
  >
    <h2
      class="pf-c-nav__section-title"
      id="group"
    >
      <div>
        group
      </div>
    </h2>
    <ul
      class="pf-c-nav__list chr-c-section-nav"
    />
  </section>
</div>
`;

exports[`ChromeNavItemFactory should render chrome nav item 1`] = `
<div>
  <li
    class="pf-c-nav__item"
    data-ouia-component-id="Foo"
    data-ouia-component-type="PF4/NavItem"
    data-ouia-safe="true"
  >
    <a
      class="pf-c-nav__link"
      data-quickstart-id="foo"
      data-testid="router-link"
      href="/foo"
    >
      Foo
       
    </a>
  </li>
</div>
`;
